src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,289 @@
+-/
+instance : group (quaternion_group n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (i | i) (j | j) (k | k);
+    simp only [mul];
+    ring,
+    simp only [add_left_inj, add_right_inj],
+    calc -(n : zmod (2 * n)) = 0 - (n : zmod (2 * n)) : by rw zero_sub
+      ... = (2 * n : zmod (2 * n)) - (n : zmod (2 * n) ) :
+        by { rw ← zmod.nat_cast_self (2 * n), norm_cast }
+      ... = (n : zmod (2 * n)) : by ring
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (i | i),
+    exact congr_arg a (zero_add i),
+    exact congr_arg xa (sub_zero i),
+  end,
+  mul_one := begin
+    rintros (i | i),
+    exact congr_arg a (add_zero i),
+    exact congr_arg xa (add_zero i),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (i | i),
+    exact congr_arg a (neg_add_self i),
+    exact congr_arg a (sub_self (n + i)),
+  end }
+
+variable {n}
+
+@[simp] lemma a_mul_a (i j : zmod (2 * n)) : a i * a j = a (i + j) := rfl
+@[simp] lemma a_mul_xa (i j : zmod (2 * n)) : a i * xa j = xa (j - i) := rfl
+@[simp] lemma xa_mul_a (i j : zmod (2 * n)) : xa i * a j = xa (i + j) := rfl
+@[simp] lemma xa_mul_xa (i j : zmod (2 * n)) : xa i * xa j = a (n + j - i) := rfl
+
+lemma one_def : (1 : quaternion_group n) = a 0 := rfl
+
+private def fintype_helper : (zmod (2 * n) ⊕ zmod (2 * n)) ≃ quaternion_group n :=
+{ inv_fun := λ i, match i with
+                 | (a j) := sum.inl j
+                 | (xa j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := a j
+                 | (sum.inr j) := xa j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/-- The special case that more or less by definition `quaternion_group 0` is isomorphic to the
+infinite dihedral group. -/
+def quaternion_group_zero_equiv_dihedral_zero : quaternion_group 0 ≃* dihedral 0 :=
+{ to_fun :=  λ i, match i with
+                | (a j) := dihedral.r j
+                | (xa j) := dihedral.sr j
+                end,
+  inv_fun := λ i, match i with
+                | (dihedral.r j) := a j
+                | (dihedral.sr j) := xa j
+                end,
+  left_inv := by rintro (k | k); refl,
+  right_inv := by rintro (k | k); refl,
+  map_mul' := begin rintros (k | k) (l | l); try {refl},
+                    change quaternion_group_zero_equiv_dihedral_zero._match_1 (a (0 + l - k)) = _,
+                    rw zero_add,
+                    refl end }
+
+/-- Some of the lemmas on `zmod m` require that `m` is positive, as `m = 2 * n` is the case relevant
+in this file but we don't want to write `[fact (0 < 2 * n)]` we make this lemma a local instance. -/
+private lemma two_mul_pos_of_pos [fact (0 < n)] : fact (0 < 2 * n) :=
+begin
+rw ← zero_mul 0,
+          apply mul_lt_mul',
+            { exact zero_le 2 },
+            { assumption },
+            { exact zero_le 0 },
+            { exact nat.zero_lt_succ 1 }
+end
```

Comment:
Please put
```
lemma two_mul_pos (h : 0 < n) : 0 < 2 * n :=
mul_pos (by norm_num) h
```
somewhere central, then use it.

Reply:
Two questions: 1) Shouldn't what we put elsewhere be generalised to 
```
lemma succ_mul_pos {m : ℕ} (h : 0 < n) : 0 < (nat.succ m) * n :=
mul_pos (by norm_num) h
```
I checked and the file also works with making the fact version of this lemma a local instance.
2) How would I then add the facts-version? Would I just restate it using the lemma put elsewhere?

Reply:
1) Yes, that generalization is good. I think you could also keep `two_mul_pos` as a separately stated consequence, so it can be used when syntactic matching matters. (But since it's not an equality it's not so important.)

2) Yes, promoting it to a local fact can be done with a private lemma.